internal/docs/generate_test.go

```diff
@@ -0,0 +1,216 @@
+package docs
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/fs"
+	"path/filepath"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/spf13/afero"
+
+	"github.com/crossplane/crossplane-runtime/pkg/test"
+)
+
+func TestIndexerRun(t *testing.T) {
+	type args struct {
+		path string
+		fs   func() afero.Fs
+	}
+
+	type want struct {
+		m   string
+		err error
+	}
+
+	cases := map[string]struct {
+		reason string
+		args   args
+		want   want
+	}{
+		"Successful": {
+			reason: "Should be able to generate json from directory structure for files",
+			args: args{
+				path: "/docs",
+				fs: func() afero.Fs {
+					fs := afero.NewMemMapFs()
+					fs = Write(t, fs, "/docs/1.md", "---\ntitle: Title\nweight: 99\n---")
+					return fs
+				},
+			},
+			want: want{
+				m: func() string {
+					b, _ := json.MarshalIndent([]Item{{
+						DisplayName: "Title",
+						Location:    "1.md",
+					}}, "", "\t")
+					return string(b)
+				}(),
+			},
+		},
+		"SuccessfulMultiLevel": {
+			reason: "Should be able to generate json from directory structure for multi-directory files",
+			args: args{
+				path: "/docs",
+				fs: func() afero.Fs {
+					fs := afero.NewMemMapFs()
+					fs = Write(t, fs, "/docs/advanced/_index.md", "---\nsection: Advanced\nweight: 1\n---")
+					fs = Write(t, fs, "/docs/advanced/1.md", "---\ntitle: Title\nweight: 99\n---")
+					fs = Write(t, fs, "/docs/1.md", "---\ntitle: Title 2\nweight: 1\n---")
+					return fs
+				},
+			},
+			want: want{
+				m: func() string {
+					b, _ := json.MarshalIndent([]Item{{
+						DisplayName: "Title 2",
+						Location:    "1.md",
+					}, {
+						DisplayName: "Advanced/Title",
+						Location:    "advanced/1.md",
+					}}, "", "\t")
+					return string(b)
+				}(),
+			},
+		},
+		"SuccessfulMultiLevelOtherStuff": {
+			reason: "Should be able to generate json from directory structure for multi-directory files with other content",
+			args: args{
+				path: "/docs",
+				fs: func() afero.Fs {
+					fs := afero.NewMemMapFs()
+					fs = Write(t, fs, "/docs/advanced/_index.md", "---\nsection: Advanced\nweight: 1\n---")
+					fs = Write(t, fs, "/docs/advanced/1.md", "---\ntitle: Title\nweight: 99\n---")
+					fs = Write(t, fs, "/docs/1.md", "---\ntitle: Title 2\nweight: 1\n---")
+					fs = Write(t, fs, "/docs/quickstart.sh", "#bash script here")
+					fs = Write(t, fs, "/docs/image.gif", "image contents")
+					return fs
+				},
+			},
+			want: want{
+				m: func() string {
+					b, _ := json.MarshalIndent([]Item{{
+						DisplayName: "Title 2",
+						Location:    "1.md",
```

Comment:
@nullable-eth @hasheddan Turns out unit tests were not passing because the tests [expected](https://github.com/upbound/official-providers/blob/main/updoc/internal/generate_test.go#L47) `/` prefix, i.e. `/1.md` instead of `1.md`. I changed the expected output to `1.md` variant since the committed JSON in providers didn't have the `/` prefix but please check.